### PR TITLE
Major Feature: Concurrent Landkid Builds

### PR DESCRIPTION
**Pull request by lukebatchelor**
* Landkid now runs multiple builds at the same time, splitting up by target branch
* Builds that are optimistically run will be sent a list of commits that they need to merge in at the start of their build
* Builds that have a dependency fail will be removed from running and will be re-added to the queue
* Fixes a ton of bugs with out of date data and race 
* Ability to remove any PR from the UI correctly
* Shows the whole sequence of status changes that a single land request may have gone through

### CODE REVIEW

File: src/db/MigrationService.ts [hunk 1]
```diff
@@ -48,39 +48,40 @@ export class MigrationService {
 }
 
 if (require.main === module) {
-  const config = require(resolve(process.cwd(), 'config.js')) as Config;
+  console.log('This script currently only works programmatically');
+  // const config = require(resolve(process.cwd(), 'config.js')) as Config;
```
Comment (lukebatchelor) [inline src/db/MigrationService.ts]: What's happening with these @gardnerjack ?

File: src/db/index.ts [hunk 2]
```diff
@@ -115,16 +160,17 @@ export class LandRequestStatus extends Model<LandRequestStatus> implements IStat
...
-  readonly reason: string | null;
+  readonly reason: string;
 
   @AllowNull(false)
   @Column(
     Sequelize.ENUM({
       values: [
         'will-queue-when-ready',
-        'created',
+        'created', // actually not used - remove
```
Comment (lukebatchelor) [inline src/db/index.ts]: follow up

File: src/lib/History.ts [hunk 3]
```diff
@@ -13,32 +13,37 @@ export class LandRequestHistory {
...
-    const allHistoryData = await LandRequest.findAll<LandRequest>({
-      where: {
-        id: {
-          $in: latestLandRequestStatuses.rows.map(s => s.requestId),
-        },
-      },
-      order: [['created', 'DESC']],
-      include: [PullRequest, LandRequestStatus],
-    });
+    // const allHistoryData = await LandRequest.findAll<LandRequest>({
```
Comment (lukebatchelor) [inline src/lib/History.ts]: Follow up, delete this

File: src/lib/Queue.ts [hunk 4]
```diff
@@ -17,12 +17,41 @@ export class LandRequestQueue {
...
+      },
+      order: [['date', 'ASC']],
+      include: [
+        {
+          model: LandRequest,
+          include: [PullRequest],
+        },
+      ],
+    });
+    const sortedQueue = [...queue].sort((a, b) => {
```
Comment (lukebatchelor) [inline src/lib/Queue.ts]: Follow up, add a comment here

File: src/static/current-state/components/QueueItem.tsx [hunk 5]
```diff
@@ -152,27 +181,103 @@ export class QueueItem extends React.Component<QueueItemProps> {
...
+        this.setState({
+          status: landRequestInfo.statuses.find(
+            (status: IStatusUpdate) => status.isLatest === true,
+          ),
+          landRequestInfo,
+        }),
+      );
+  };
+
+  renderMoreInfo = (status: IStatusUpdate, dependsOn: string[], bitbucketBaseUrl: string) => {
```
Comment (lukebatchelor) [inline src/static/current-state/components/QueueItem.tsx]: I hate that I'm passing the baseUrl around like this, but 🤷‍♂

File: typings/ambient.d.ts [hunk 6]
```diff
@@ -81,13 +89,13 @@ declare namespace Express {
   }
 }
 
-declare interface HistoryItem {
-  request: ILandRequest;
-  statusEvents: Array<IStatusUpdate>;
-}
+// declare interface HistoryItem {
```
Comment (lukebatchelor) [inline typings/ambient.d.ts]: follow up, remove

File: src/db/MigrationService.ts (hunk 1 above)
Comment (gardnerjack) [reply src/db/MigrationService.ts]: It's broken. Michael has a similar issue with his so might see if he found a solution.

File: src/lib/Runner.ts [hunk 7]
```diff
@@ -286,15 +351,15 @@ export class Runner {
...
-    const landRequestInfo = await this.queue.maybeGetStatusForQueuedRequestById(requestId);
-    if (!landRequestInfo) return false;
+  removeLandRequestFromQueue = async (requestId: string, user: ISessionUser): Promise<boolean> => {
+    const landRequestStatus = await this.queue.maybeGetStatusForQueuedRequestById(requestId);
+    if (!landRequestStatus) return false;
 
-    await landRequestInfo.request.setStatus(
+    await landRequestStatus.request.setStatus(
       'aborted',
       `Removed from queue by user "${user.aaid}" (${user.displayName})`,
```
Comment (gardnerjack) [inline src/lib/Runner.ts]: Only displayName if it exists, otherwise aaid

**Pull request merged**